src/librustc/ty/mod.rs

```diff
@@ -1034,7 +1034,7 @@ impl<'tcx> ToPolyTraitRef<'tcx> for PolyProjectionPredicate<'tcx> {
         // This is because here `self` has a `Binder` and so does our
         // return value, so we are preserving the number of binding
         // levels.
-        ty::Binder(self.0.projection_ty.trait_ref)
+        ty::Binder(ty::tls::with(|tcx| self.0.projection_ty.trait_ref(tcx)))
```

Comment:
I agree with @eddyb that it is what dubious for this type to implement `to_poly_trait_ref`. An inherent function seems better. Basically is converting a predicate like `T: Iterator<Item=Foo>` into the underlying `T: Iterator` trait-reference -- that's not *wrong*, but it feels like it merits an explicit "cast". This was (however) equally true before, I guess -- but still better to fix.